apps/Gruntfile.js

```diff
@@ -706,22 +707,100 @@ describe('entry tests', () => {
+        // shared modules. This would mean that state within webpack modules
+        // cannot be shared between entry points, breaking many assumptions made
+        // by our application. For more information, see:
+        // https://webpack.js.org/concepts/manifest/#runtime
+        // https://webpack.js.org/configuration/optimization/#optimizationruntimechunk
+        //
+        // In the future, if we can limit ourselves to one webpack entry point
+        // per page, we could consider removing the runtimeChunk config.
+        runtimeChunk: {
+          name: 'webpack-runtime'
+        },
+        splitChunks: {
+          cacheGroups: {
+            // Pull any module shared by 2+ appsEntries into the "common" chunk.
+            common: {
+              name: 'common',
+              minChunks: 2,
+              chunks: chunk => {
+                return _.keys(appsEntries).includes(chunk.name);
+              }
+            },
+            // Pull any module shared by 2+ codeStudioEntries into the
+            // "code-studio-common" chunk.
+            'code-studio-common': {
+              name: 'code-studio-common',
+              minChunks: 2,
+              chunks: chunk => {
+                const chunkNames = _.keys(codeStudioEntries);
+                return chunkNames.includes(chunk.name);
+              },
+              priority: 10
+            },
+            // With just the cacheGroups listed above, we end up with many
+            // duplicate modules between the "common" and "code-studio-common"
+            // chunks. The next cache group eliminates some of this duplication
+            // by pulling more modules from "common" into "code-studio-common".
+            //
+            // The cacheGroup below opportunistically pulls modules from
+            // "common" into "code-studio-common". The key is that it uses
+            // minChunks to prevent any modules used only by appsEntries from
+            // being extracted, providing a guarantee that we don't
+            // unnecessarily increase the download size for pages which include
+            // code-studio-common.js but not common.js.
+            //
+            // There is no converse guarantee that this strategy will eliminate
+            // all duplication between "common" and "code-studio-common".
+            // However, at the time of this writing, bundle analysis indicates
+            // that is currently effective in eliminating any duplication.
+            //
+            // In the future, we want to move toward asynchronous imports, which
+            // allow webpack to manage bundle splitting and sharing behind the
+            // scene. Once we adopt this approach, the need for predefined
+            // cacheGroups will go away.
+            //
+            // For more information see: https://webpack.js.org/guides/code-splitting/
+            'code-studio-multi': {
```

Comment:
In webpack 3, we "chained" our chunks config, in the sense that
the "code-studio-common" chunk was able to list the "common"
chunk (the result of a previous chunking operation) as one of its
inputs. This type of chaining behavior is no longer possible in
webpack 4, so I had to get creative to avoid a regression in chunk sizes.